### ISSUE TITLE
Serializing to Turtle seems off

### ISSUE THREAD

**Issue by b-zee** [linkeddata/rdflib.js#251]
When serializing a graph to Turtle, it seems there is an extra unused prefix.

Consider the following snippet:
```
const rdf = require('rdflib');

const graph = rdf.graph();
const FOAF = rdf.Namespace('http://xmlns.com/foaf/0.1/');

graph.add(rdf.sym('http://me.com/'), FOAF('knows'), rdf.sym('http://him.com'));
graph.add(rdf.sym('http://me.com/'), FOAF('knows'), rdf.sym('http://her.com'));

const serial = rdf.serialize(undefined, graph, 'http://me.com/', 'text/turtle');
console.log(serial);
```

This produces:
```
@prefix : </#>.
@prefix me: <>.
@prefix n0: <http://xmlns.com/foaf/0.1/>.

me: n0:knows <http://her.com>, <http://him.com>.
```

The empty `:` prefix is not used.

Also, I'm missing the `@base` or any reference to the main subject (`http://me.com/`).

**Comment by kjetilk** [comment]
Yes, and this highlights an issue that have been on my mind for some time. The Turtle produced by the serializer isn't very readable and easy to understand intuitively, and that is one of Turtle's strengths.

It may seem like a small thing, but I believe it is not. The Turtle must be very readable, so that newcomers can read it intuitively.

We must also make sure prefixes are good. Over in the Perl world, we have three modules for making sensible qnames: [RDF::Prefixes](https://metacpan.org/pod/RDF::Prefixes) which tries to guess prefixes by looking at the URI, [RDF::NS](https://metacpan.org/pod/RDF::NS) which uses http://prefix.cc/ which is a crowdsourced list of prefix-URI pairs, and [RDF::NS::Curated](https://metacpan.org/pod/RDF::NS::Curated) which is a curated such set. [URI::NamespaceMap](https://metacpan.org/pod/URI::NamespaceMap) is used to guess and manage prefix-URI pairs. Perhaps we should do something similar to ensure that we always have good qnames?

Then, I think we should think about serializing with some assumptions that are not actually a part of the data model if possible. Like, we use qnames for all the "vocabulary terms", but not for resources that we talk about. For that, we should use `@base` and pointy brackes. Perhaps there should be a bit more logic in the serializer, but I think that would help a lot.

**Comment by ManuelTS** [comment]
Any update from this? I'm having the exact same issue with the two `@prefixe`s and a missing `@base`

**Comment by kjetilk** [comment]
I'm not aware of anything, I'm afraid.

**Comment by jg10-mastodon-social** [comment]
If I understand the code correctly, the default prefix can be suppressed by adding the d flag.

https://github.com/linkeddata/rdflib.js/blob/3fa7742edc0f9b5ce68c10a22faa15da819be3f5/src/serializer.js#L529-L533

The same `prefixDirectivesMethod` function is probably where the base could be added, perhaps also guarded behind a flag.
